### DISCUSSION TITLE
Accessing cache from onError / onErrorRetry

### DISCUSSION THREAD

**Discussion by joshkel** [vercel/swr#1755] [Ideas]
I'd like to access the cache from an `onError` handler so that I can invalidate previously cached values in response to certain types of errors.

I see that `onError` takes the `config` object as its third parameter, and in practice, the `config` object appears to contain a `cache` member. However, it doesn't appear to be permitted by the TypeScript types or documented.

Is it permitted to access this member? Can/should the types be updated to reflect this? Or can/should SWR add an alternative standardized way of accessing the cache for the current request?

Thank you.

**Comment by shuding** [comment]
> so that I can invalidate previously cached values in response to certain types of errors.

I think in that case, I wouldn’t directly edit the cache, because it won’t sync that state with all hooks. Maybe we should expose `mutate` in those callbacks.

**Reply by joshkel** [reply]
Thanks for the reply. In my ideal world, I'd actually want to access both the cache and `mutate`, because I'm selectively mutating based on whether previous values had succeeded. (See https://github.com/vercel/swr/discussions/998.)

Or, if there were a way to call `mutate` and set both `data` and `error`, that would be better. For example, if ScopedMutator were overloaded:

```ts
(key: Key, data?: Data | Promise<Data> | MutatorCallback<Data>, shouldRevalidate?: boolean): Promise<Data | undefined>;
// To facilitate overloading, assume / require that any errors are of Error type.
// Promise and callback are not implemented, since data + error should be provided together:
(key: Key, data?: Data, error?: Error, shouldRevalidate?: boolean): Promise<Data | undefined>;
```

Then my error handler from https://github.com/vercel/swr/discussions/998 could be

```ts
function onError(err: Error, key: string, config) {
  if (err instanceof BackendError && err.response.statusCode === StatusCodes.NOT_FOUND) {
    // If the requested entity doesn't exist, clear any cached value and return the Not Found error.
    config.mutate(key, null, err, false);
  }
}
```

(Alternatively, the error object could be an optional fourth parameter to `mutate`, to avoid overloading. Or, to keep full support for callbacks and promises, the API could be extended to offer an optional flag or symbol for providing a `{data, error}`.)

**Comment by miso-belica** [comment]
Hello, I have a similar need to access the cache in `onError` but I don't have to mutate them. My use-case is as follows. I report all errors globally in `onError` to a monitoring service. But there is quite a lot of `NetworkError` which I can ignore if data are already in the cache. Problem is I want to report cases when data are not in the cache but network error happens. Currently I can't if I don't want to rely on private `cache` property. Any other idea how to do it?
